### PR TITLE
Add verify-spelling.sh for ja content

### PR DESCRIPTION
**Pull request by oomichi**
We have the translation guideline as [1], and that contains what are valid Japanese words for the corresponding English words. This adds a script to verify those words based on the guideline.

[1]: https://kubernetes.io/ja/docs/contribute/localization/

### CODE REVIEW

**Comment by oomichi** [comment]
/language ja
/cc @t-inu 


Please check this script.

File: scripts/ja/verify-spelling.sh
Comment (oomichi) [inline scripts/ja/verify-spelling.sh:25]: We could add more rules, but I'd like to see this is acceptable or not before taking more effort.

**Comment by atoato88** [comment]
FYI:
Here is the rules for translating to Japanese.
https://kubernetes.io/ja/docs/contribute/localization/#%E9%A0%BB%E5%87%BA%E5%8D%98%E8%AA%9E

And I'll comment about some other rules refered from that page.

File: scripts/ja/verify-spelling.sh
Comment (atoato88) [inline scripts/ja/verify-spelling.sh:25]:
```suggestion
TARGET_WORDS="${TARGET_WORDS} アーキテクチャー:アーキテクチャ"
TARGET_WORDS="${TARGET_WORDS} コントローラ:コントローラー"
```

I'll add one rule related to "コントローラー" at here refered from [Japanese translation rules](https://kubernetes.io/ja/docs/contribute/localization/#%E9%A0%BB%E5%87%BA%E5%8D%98%E8%AA%9E).

Invalid: コントローラ
valid: コントローラー
It's an opposite rule to other rules.

We should match just "コントローラ", so I add `-w` parameter to grep.

File: scripts/ja/verify-spelling.sh
Comment (atoato88) [inline scripts/ja/verify-spelling.sh:33]:
```suggestion
	grep "${invalid}" * -R -w
```

**Review by atoato88 — COMMENTED** [review]
I left some comments about matching the word as-is.

File: scripts/ja/verify-spelling.sh
Comment (oomichi) [reply scripts/ja/verify-spelling.sh:25]: hmm, that doesn't work on my side:
```
$ git diff
diff --git a/scripts/ja/verify-spelling.sh b/scripts/ja/verify-spelling.sh
index fa1486b..12e1b0e 100755
--- a/scripts/ja/verify-spelling.sh
+++ b/scripts/ja/verify-spelling.sh
@@ -23,6 +23,7 @@ cd "${ROOT}/${CONTENT_DIR}"
 TARGET_WORDS=""               # invalid:valid
 TARGET_WORDS="${TARGET_WORDS} コンテナー:コンテナ"
 TARGET_WORDS="${TARGET_WORDS} アーキテクチャー:アーキテクチャ"
+TARGET_WORDS="${TARGET_WORDS} コントローラ:コントローラー"
 
 INVALID_CONTAINED=""
 
@@ -30,7 +31,7 @@ for word_set in $(echo ${TARGET_WORDS})
 do
        invalid=$(echo ${word_set} | awk -F: '{print $1}')
        valid=$(echo ${word_set} | awk -F: '{print $2}')
-       grep "${invalid}" * -R
+       grep "${invalid}" * -R -w
        if [ $? -eq 0 ]; then
                echo "Invalid word ${invalid} is contained under ${CONTENT_DIR}."
                echo "Please replace ${invalid} with ${valid}."
```
After reverting the change, the original script detects many invalid words like
```
$ git checkout -f
$ ./scripts/ja/verify-spelling.sh 
docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:In-treeボリュームプラグインに関連付けられたコードは、コアKubernetesコードベースの一部として提供されます。In-treeボリュームプラグインのデプロイでは、追加のスクリプトをインストールしたり、個別のコンテナ化されたプラグインコンポーネントをデプロイしたりする必要はありません。これらのプラグインは、ストレージバックエンドでのボリュームのプロビジョニング/プロビジョニング解除とサイズ変更、Kubernetesノードへのボリュームのアタッチ/アタッチ解除、Pod内の個々のコンテナーへのボリュームのマウント/マウント解除を処理できます。次のIn-treeプラグインは、Windowsノードをサポートしています。:
docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:| オーバーレイ(KubernetesのWindows用のオーバーレイネットワークは *アルファ* 段階です) | コンテナには、外部のvSwitchに接続されたvNICが付与されます。各オーバーレイネットワークは、カスタムIPプレフィックスで定義された独自のIPサブネットを取得します。オーバーレイネットワークドライバーは、VXLANを使用してカプセル化します。 | 外部ヘッダーでカプセル化されます。 | [Win-overlay](https://github.com/containernetworking/plugins/tree/master/plugins/main/windows/win-overlay)、Flannel VXLAN (win-overlayを使用) | win-overlayは、仮想コンテナーネットワークをホストのアンダーレイから分離する必要がある場合に使用する必要があります(セキュリティ上の理由など)。データセンター内のIPが制限されている場合に、(異なるVNIDタグを持つ)異なるオーバーレイネットワークでIPを再利用できるようにします。このオプションには、Windows Server 2019で[KB4489899](https://support.microsoft.com/help/4489899)が必要です。|
docs/reference/_index.md:* [kubelet](/docs/reference/command-line-tools-reference/kubelet/) - 各ノード上で動作する最も重要なノードエージェントです。kubeletは一通りのPodSpecを受け取り、コンテナーが実行中で正常であることを確認します。
docs/tasks/run-application/run-replicated-stateful-application.md:`mysql:5.7`イメージを使って一時的なコンテナーを実行し、`mysql`クライアントバイナリーを実行します。
docs/concepts/architecture/garbage-collection.md:{{<glossary_tooltip text="kubelet" term_id="kubelet">}}は未使用のイメージに対して5分ごとに、未使用のコンテナーに対して1分ごとにガベージコレクションを実行します。
docs/concepts/architecture/garbage-collection.md:未使用のコンテナーとイメージのガベージコレクションのオプションを設定するには、[設定ファイル](/docs/tasks/administer-cluster/kubelet-config-file/)を使用してkubeletを調整し、[`KubeletConfiguration`](/docs/reference/config-api/kubelet-config.v1beta1/#kubelet-config-k8s-io-v1beta1-KubeletConfiguration)リソースタイプを使用してガベージコレクションに関連するパラメーターを変更します。
docs/concepts/architecture/garbage-collection.md:  * `MaxContainers`: クラスターが持つことができるデッドコンテナーの最大数。`0`未満に設定すると無効化されます。
docs/concepts/architecture/garbage-collection.md:`MaxPerPodContainer`と`MaxContainers`は、Podごとのコンテナーの最大数(`MaxPerPodContainer`)を保持すると、グローバルなデッドコンテナの許容合計(`MaxContainers`)を超える状況で、互いに競合する可能性があります。
docs/concepts/architecture/garbage-collection.md:この状況では、kubeletは`MaxPerPodContainer`を調整して競合に対処します。最悪のシナリオは、`MaxPerPodContainer`を1にダウングレードし、最も古いコンテナーを削除することです。
docs/concepts/configuration/manage-resources-containers.md:そのコンテナに4GiBの`メモリー`制限を設定すると、kubelet(および{{< glossary_tooltip text="コンテナランタイム" term_id="container-runtime" >}}) が制限を適用します。ランタイムは、コンテナーが設定済みのリソース制限を超えて使用するのを防ぎます。例えば、コンテナ内のプロセスが、許容量を超えるメモリを消費しようとすると、システムカーネルは、メモリ不足(OOM)エラーで、割り当てを試みたプロセスを終了します。
docs/concepts/configuration/manage-resources-containers.md:制限は、違反が検出されるとシステムが介入するように事後的に、またはコンテナーが制限を超えないようにシステムが防ぐように強制的に、実装できます。
docs/concepts/configuration/secret.md:コンテナーはSecretのデータをSSH接続を確立するために使用することができます。
docs/concepts/storage/persistent-volumes.md:Podにrawブロックデバイスを追加する場合は、マウントパスの代わりにコンテナーでデバイスパスを指定します。
docs/concepts/workloads/controllers/deployment.md:Deploymentのロールアウトは、DeploymentのPodテンプレート(この場合`.spec.template`)が変更された場合にのみトリガーされます。例えばテンプレートのラベルもしくはコンテナーイメージが更新された場合です。Deploymentのスケールのような更新では、ロールアウトはトリガーされません。
docs/concepts/workloads/controllers/deployment.md:設定ファイルの利用に関する情報は[アプリケーションのデプロイ](/ja/docs/tasks/run-application/run-stateless-application-deployment/)を参照してください。コンテナーの設定に関しては[リソースを管理するためのkubectlの使用](/ja/docs/concepts/overview/working-with-objects/object-management/)を参照してください。
docs/concepts/workloads/controllers/deployment.md:`.spec.minReadySeconds`はオプションのフィールドで、新しく作成されたPodが利用可能となるために、最低どれくらいの秒数コンテナーがクラッシュすることなく稼働し続ければよいかを指定するものです。デフォルトでは0です(Podは作成されるとすぐに利用可能と判断されます)。Podが利用可能と判断された場合についてさらに学ぶために[Container Probes](/ja/docs/concepts/workloads/pods/pod-lifecycle/#container-probes)を参照してください。
Invalid word コンテナー is contained under content/ja.
Please replace コンテナー with コンテナ.

docs/tasks/administer-cluster/nodelocaldns.md:NodeLocal DNSキャッシュは、クラスターノード上でDNSキャッシュエージェントをDaemonSetで稼働させることで、クラスターのDNSパフォーマンスを向上させます。現在のアーキテクチャーにおいて、ClusterFirstのDNSモードでのPodは、DNSクエリー用にkube-dnsのService IPに疎通します。これにより、kube-proxyによって追加されたiptablesを介してkube-dns/CoreDNSのエンドポイントへ変換されます。この新しいアーキテクチャーによって、Podは同じノード上で稼働するDNSキャッシュエージェントに対して疎通し、それによってiptablesのDNATルールとコネクショントラッキングを回避します。ローカルのキャッシュエージェントはクラスターのホスト名(デフォルトではcluster.localというサフィックス)に対するキャッシュミスがあるときはkube-dnsサービスへ問い合わせます。
docs/tasks/administer-cluster/nodelocaldns.md:* 現在のDNSアーキテクチャーでは、ローカルのkube-dns/CoreDNSがないとき、DNSへの秒間クエリー数が最も高いPodは他のノードへ疎通する可能性があります。ローカルでキャッシュを持つことにより、この状況におけるレイテンシーの改善に役立ちます。
docs/tasks/administer-cluster/nodelocaldns.md:## アーキテクチャー図
Invalid word アーキテクチャー is contained under content/ja.
Please replace アーキテクチャー with アーキテクチャ.
```

Comment (oomichi) [reply scripts/ja/verify-spelling.sh:25]: `-w` option of `grep` command like
```
       -w, --word-regexp
              Select  only  those  lines  containing  matches  that form whole words.  The test is that the matching substring must either be at the beginning of the line, or preceded by a non-word
              constituent character.  Similarly, it must be either at the end of the line or followed by a non-word constituent character.  Word-constituent characters are letters, digits, and  the
              underscore.
```
Those words are written without any spaces around the words, then I guess that doesn't work.

Comment (atoato88) [reply scripts/ja/verify-spelling.sh:25]: I checked these commands on terminal, and confirmed that works well with "コントローラ", but desn't work with "コンテナー".
"grep" do something special with the word trailing "ー(長音符)"...?
```bash
# it works in spite of without space in the both side on コントローラ
~/go/src/k8s.io/website $ grep "コントローラ" * -R -w
content/ja/docs/tasks/debug/debug-cluster/_index.md:  - サポートサービス(ノードコントローラ、レプリケーションコントローラーマネージャー、スケジューラーなど)VMのシャットダウンまたはクラッシュ
scripts/ja/verify-spelling.sh:TARGET_WORDS="${TARGET_WORDS} コントローラ:コントローラー"

# it doesn't work..
~/go/src/k8s.io/website $ grep "コンテナー" * -R -w
scripts/ja/verify-spelling.sh:TARGET_WORDS="${TARGET_WORDS} コンテナー:コンテナ"
```

I think there are some word trail "ー(長音符)" is valid, on the other hand word trail "ー(長音符)" is invalid in Japanese.(It's so confusing..)
We should support both case, but as of now this is the corner case, so ignore my suggestion.

I will check and create other update in my free time.

Comment (atoato88) [inline scripts/ja/verify-spelling.sh:25]:
```suggestion
TARGET_WORDS="${TARGET_WORDS} アーキテクチャー:アーキテクチャ"
TARGET_WORDS="${TARGET_WORDS} コントローラ[^ー]:コントローラー"
```

File: scripts/ja/verify-spelling.sh
Comment (atoato88) [inline scripts/ja/verify-spelling.sh:33]:
```suggestion
	grep -E "${invalid}" * -R
```

**Review by atoato88 — COMMENTED** [review]
Just FYI.
This is an other suggestions by using `-E` parameter on `grep`.
This works fine, but we should add `TARGET_WORDS` to regex pattern like `コントローラ[^ー]` on invalid pattern.

**Comment by oomichi** [comment]
@atoato88 Thanks for your comments, I will try your suggention later.
BTW do you have any idea where is the best place to hold this kind of language specific scripts?
I feel it is better to put those scripts under `ja` directory where `ja` approver can approve.

**Comment by atoato88** [comment]
Agree.
I think it's better to put `verify-spelling.sh` on `scripts/ja/`, and also new `OWNERS` on `scripts/ja/` which includes `ja` approvers and reviewers.

If there are same situation on other languages, they can notice our `ja` directory and create some assets theirown.

**Comment by oomichi** [comment]
> Agree. I think it's better to put `verify-spelling.sh` on `scripts/ja/`, and also new `OWNERS` on `scripts/ja/` which includes `ja` approvers and reviewers.
> 
> If there are same situation on other languages, they can notice our `ja` directory and create some assets theirown.

Thank you for the suggestion, that seems a good idea for me!
Let me try to add the OWNERS file in this pull request.

File: scripts/ja/verify-spelling.sh
Comment (oomichi) [reply scripts/ja/verify-spelling.sh]: Thanks, that works!

**Comment by oomichi** [comment]
@atoato88 I applied your suggestion(`-E` option) and confimred that works!
Do you have any more points at this time?

**Comment by atoato88** [comment]
@oomichi 
Thank you for quick response!
I don't have any points in this time :+1: 

We have other points followings:
- add other `TARGET_WORD`
- update output text more human friendly

But now, we prefer to go forward and these items are low priority, I think.

**Review by inductor — APPROVED** [review]
nice!

**Comment by oomichi** [comment]
@tengqm 
Hello, could you take a look at this pull request?
This is a script for `ja` language and the `ja` approvers put `lgtm` on this already.

**Comment by atoato88** [comment]
@bradtopol @tengqm 
Gentle ping.
Could you reveiw this PR?

**Comment by oomichi** [comment]
just rebased for the latest main branch.

**Comment by sftim** [comment]
I recommend updating the [README](https://github.com/kubernetes/website/tree/bb2aa4281681daacd8339d687c01d7307721596d/scripts) for scripts to mention the scripts in the `ja` directory (perhaps also the `releng` tooling, which I think right now doesn't get mentioned either).

File: scripts/ja/verify-spelling.sh
Comment (atoato88) [inline scripts/ja/verify-spelling.sh:26]:
```suggestion
TARGET_WORDS="${TARGET_WORDS} コントローラ[^ー]:コントローラー"
TARGET_WORDS="${TARGET_WORDS} オペレータ[^ー]:オペレーター"
TARGET_WORDS="${TARGET_WORDS} パラメータ[^ー]:パラメーター"
```

Recently, I reviewed [the PR](https://github.com/kubernetes/website/pull/38185#pullrequestreview-1213656718) what this script should find in.
I would like to add the words `オペレーター` and `パラメーター` to this PR now, because those are commented on [that PR](https://github.com/kubernetes/website/pull/38185#pullrequestreview-1213656718).

**Review by atoato88 — COMMENTED** [review]
I added the additional expressions, because I would like to use this script in [other PR](https://github.com/kubernetes/website/pull/38185#pullrequestreview-1213656718).

**Comment by oomichi** [comment]
> I recommend updating the [README](https://github.com/kubernetes/website/tree/bb2aa4281681daacd8339d687c01d7307721596d/scripts) for scripts to mention the scripts in the `ja` directory (perhaps also the `releng` tooling, which I think right now doesn't get mentioned either).

@sftim Thanks for your review.
That makes it better, I updated the README to add the script description.

File: scripts/ja/verify-spelling.sh
Comment (oomichi) [reply scripts/ja/verify-spelling.sh:26]: Thank you, that is a nice point.
I have added those lines.

**Comment by atoato88** [comment]
/lgtm
Thank you for quick update!

File: scripts/README.md
Comment (nasa9084) [inline scripts/README.md:185]: maybe sometimes they're not "invalid" though

suggest:
```suggestion
This script finds Japanese words that are against the guideline[1]
```

Comment (oomichi) [reply scripts/README.md:185]: Thanks for your review @nasa9084 
That makes sense, I applied the suggestion on the latest pr.

**Review by nasa9084 — APPROVED** [review]

**Comment by oomichi** [comment]
@sftim Could you take a look at this again?

**Comment by oomichi** [comment]
@divya-mohan0209 Could you take a look at this ?

**Comment by atoato88** [comment]
@sftim
gentle ping.

File: scripts/README.md
Comment (kbhawkey) [inline scripts/README.md:15]: nit: It's not clear what the 'guideline' is referring to.

File: scripts/ja/OWNERS
Comment (kbhawkey) [inline scripts/ja/OWNERS:2]: Is there a good reason why you need an OWNERS file for the script?

File: scripts/README.md
Comment (oomichi) [reply scripts/README.md:15]: The guideline is https://kubernetes.io/ja/docs/contribute/localization/

File: scripts/ja/OWNERS
Comment (oomichi) [reply scripts/ja/OWNERS:2]: Because the script is for `ja` content as the directory name.
I think it is good to get reviews from `ja` reviewers for long-term maintenance.

Comment (atoato88) [reply scripts/ja/OWNERS:2]: I agree with oomichi's comment, `ja` reviewers can review the content effectively I think.

Comment (kbhawkey) [reply scripts/ja/OWNERS:2]: Yes, I understand that you want to manage the changes to the script (which is okay), though I am cautious about adding more OWNERS files to the repo.

Comment (oomichi) [reply scripts/ja/OWNERS:2]: > Yes, I understand that you want to manage the changes to the script (which is okay)

ah, I see your point.
TBH I am not a member in sig-docs-ja-reviews/owners at all.
I just heard there is some issue of notation fluctuation in `ja` content and created this script to block it as possible to read consistent `ja` document.
I believe sig-docs-ja-reviews/owners will maintain this script to improve the content, and I will do that when necessary.

Anyways thank you for your review @kbhawkey

**Comment by nasa9084** [comment]
@bradtopol can you review this (and approve if there's no problem)?

**Pull request merged**
